src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -1135,34 +1041,47 @@ public CachedConnection(HttpConnection connection)
         /// into the waiter as a result, and if no connection is available from the pool,
         /// this waiter's logic is used to create the connection.
         /// </summary>
-        private class ConnectionWaiter : TaskCompletionSource<(HttpConnectionBase connection, bool isNewConnection, HttpResponseMessage failureResponse)>
+        private sealed class ConnectionWaiter
         {
-            /// <summary>The pool with which this waiter is associated.</summary>
-            internal readonly HttpConnectionPool _pool;
-            /// <summary>Request to use to create the connection.</summary>
-            private readonly HttpRequestMessage _request;
-
-            /// <summary>Cancellation token for the waiter.</summary>
-            internal readonly CancellationToken _cancellationToken;
-            /// <summary>Registration that removes the waiter from the registration list.</summary>
-            internal CancellationTokenRegistration _cancellationTokenRegistration;
+            private readonly TaskCompletionSource<HttpConnection> _tcs;
+
             /// <summary>Next waiter in the list.</summary>
             internal ConnectionWaiter _next;
             /// <summary>Previous waiter in the list.</summary>
             internal ConnectionWaiter _prev;
 
             /// <summary>Initializes the waiter.</summary>
-            public ConnectionWaiter(HttpConnectionPool pool, HttpRequestMessage request, CancellationToken cancellationToken) : base(TaskCreationOptions.RunContinuationsAsynchronously)
+            public ConnectionWaiter()
             {
-                Debug.Assert(pool != null, "Expected non-null pool");
-                _pool = pool;
-                _request = request;
-                _cancellationToken = cancellationToken;
+                _tcs = new TaskCompletionSource<HttpConnection>(TaskCreationOptions.RunContinuationsAsynchronously);
             }
 
-            /// <summary>Creates a connection.</summary>
-            public ValueTask<(HttpConnection, HttpResponseMessage)> CreateConnectionAsync() =>
-                _pool.CreateHttp11ConnectionAsync(_request, _cancellationToken);
+            public async ValueTask<(HttpConnectionBase connection, bool isNewConnection, HttpResponseMessage failureResponse)> 
+                GetConnectionAsync(HttpConnectionPool pool, HttpRequestMessage request, CancellationToken cancellationToken)
+            {
+                CancellationTokenRegistration cancellationTokenRegistration = default;
+                if (cancellationToken.CanBeCanceled)
+                {
+                    cancellationTokenRegistration = cancellationToken.Register(() => _tcs.TrySetCanceled());
+                }
+
+                HttpConnection connection = await _tcs.Task;
+
+                cancellationTokenRegistration.Dispose();
```

Comment:
Once you simplify the earlier stuff, this can just use a using, e.g.
```C#
using (cancellationToken.Register(s => ((TaskCompletionSource<HttpConnection>)s).TrySetCanceled(), _tcs)
{
    await _tcs.Task.ConfigureAwait(false);
}
```